File: ai-factory-backend/src/main/java/com/aifactory/entity/Foreshadowing.java
package com.aifactory.entity;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import lombok.Data;

import java.time.LocalDateTime;

/**
 * 伏笔实体
 *
 * @Author CaiZy
 * @Date 2025-01-22
 * @Wechat DevilJieH
 * @Email a314170122@outlook.com
 */
@Data
@TableName("novel_foreshadowing")
public class Foreshadowing {

    /**
     * 伏笔ID（自增主键）
     */
    @TableId(type = IdType.AUTO)
    private Long id;

    /**
     * 项目ID
     */
    private Long projectId;

    /**
     * 伏笔标题
     */
    private String title;

    /**
     * 伏笔类型：character(人物)/item(物品)/event(事件)/secret(秘密)
     */
    private String type;

    /**
     * 伏笔描述
     */
    private String description;

    /**
     * 布局类型：bright1(明线1)/bright2(明线2)/bright3(明线3)/dark(暗线)
     */
    private String layoutType;

    /**
     * 埋伏笔的章节
     */
    private Integer plantedChapter;

    /**
     * 计划填坑的章节
     */
    private Integer plannedCallbackChapter;

    /**
     * 实际填坑的章节
     */
    private Integer actualCallbackChapter;

    /**
     * 状态：pending(未填回)/in_progress(进行中)/completed(已填回)
     */
    private String status;

    /**
     * 优先级，数值越大优先级越高
     */
    private Integer priority;

    /**
     * 备注信息
     */
    private String notes;

    /**
     * 创建时间
     */
    private LocalDateTime createTime;

    /**
     * 更新时间
     */
    private LocalDateTime updateTime;
}

File: ai-factory-frontend/src/types/foreshadowing.ts
/**
 * 伏笔相关类型定义
 * Maps from backend ForeshadowingDto, ForeshadowingCreateDto, ForeshadowingUpdateDto, ForeshadowingQueryDto
 */

/**
 * 伏笔类型
 */
export type ForeshadowingType = 'character' | 'item' | 'event' | 'secret'

/**
 * 布局线类型
 */
export type ForeshadowingLayoutType = 'bright1' | 'bright2' | 'bright3' | 'dark'

/**
 * 伏笔状态
 */
export type ForeshadowingStatus = 'pending' | 'in_progress' | 'completed'

/**
 * 伏笔信息 (maps from ForeshadowingDto)
 */
export interface Foreshadowing {
  id: number
  projectId: number
  title: string
  type: ForeshadowingType
  description: string
  layoutType: ForeshadowingLayoutType | null
  plantedChapter: number
  plantedVolume: number | null
  plannedCallbackChapter: number | null
  plannedCallbackVolume: number | null
  actualCallbackChapter: number | null
  status: ForeshadowingStatus
  priority: number
  notes: string | null
  createTime: string
  updateTime: string
}

/**
 * 伏笔创建请求 (maps from ForeshadowingCreateDto)
 * Per D-05: plantedChapter is required for creation
 */
export interface ForeshadowingCreateRequest {
  title: string
  type: ForeshadowingType
  description: string
  layoutType?: ForeshadowingLayoutType
  plantedChapter: number
  plantedVolume?: number
  plannedCallbackChapter?: number
  plannedCallbackVolume?: number
  priority?: number
  notes?: string
}

/**
 * 伏笔更新请求 (maps from ForeshadowingUpdateDto)
 * Per D-05: plantedChapter is NOT included (immutable after creation)
 */
export interface ForeshadowingUpdateRequest {
  title: string
  type: ForeshadowingType
  description: string
  layoutType?: ForeshadowingLayoutType
  plannedCallbackChapter?: number
  plannedCallbackVolume?: number
  actualCallbackChapter?: number
  status?: ForeshadowingStatus
  priority?: number
  notes?: string
}

/**
 * 伏笔查询参数 (maps from ForeshadowingQueryDto)
 */
export interface ForeshadowingQueryParams {
  type?: ForeshadowingType
  layoutType?: ForeshadowingLayoutType
  status?: ForeshadowingStatus
  currentChapter?: number
  plantedChapter?: number
  plannedCallbackChapter?: number
  plantedVolume?: number
  plannedCallbackVolume?: number
}
